File: src/types/compiler.ts
/**
 * Shared compiler data models.
 * Source of truth: SystemDesign.md Section 3.
 *
 * These types are used by the mock adapter (v1), the future HTTP adapter
 * (v2, see Architecture.md Section 8), and every visualization component.
 * Never redefine these shapes locally in a feature/component file.
 */

// ---- Lexer ----
export type TokenType =
  | 'KEYWORD'
  | 'IDENTIFIER'
  | 'OPERATOR'
  | 'LITERAL'
  | 'PUNCTUATION'
  | 'COMMENT';

export interface Token {
  id: string;
  type: TokenType;
  value: string;
  line: number;
  column: number;
}

// ---- Parser ----
export interface ASTNode {
  id: string;
  kind: string; // "BinaryExpr", "IfStatement", "FunctionDecl", ...
  children: ASTNode[];
  line: number;
  metadata?: Record<string, unknown>;
}

// ---- Semantic Analysis ----
export interface SymbolEntry {
  name: string;
  type: string; // int, float, function, etc.
  scope: string;
  declaredAt: number; // line number
}

export type CompilerPhase =
  | 'lexical'
  | 'syntax'
  | 'semantic'
  | 'intermediate'
  | 'optimization'
  | 'codegen';

export interface SemanticDiagnostic {
  severity: 'error' | 'warning';
  message: string;
  line: number;
  phase: CompilerPhase;
}

// ---- Intermediate Code ----
export interface TACInstruction {
  id: string;
  op: string; // "=", "+", "goto", "if", ...
  arg1?: string;
  arg2?: string;
  result?: string;
  label?: string;
}

// ---- Optimization ----
export interface OptimizationDiff {
  before: TACInstruction[];
  after: TACInstruction[];
  passesApplied: string[]; // ["Constant Folding", "Dead Code Elimination"]
}

// ---- Target Code ----
export interface AssemblyLine {
  instruction: string;
  operands: string[];
  comment?: string;
}

// ---- Aggregate Result ----
export interface CompilationResult {
  tokens: Token[];
  ast: ASTNode | null;
  symbolTable: SymbolEntry[];
  diagnostics: SemanticDiagnostic[];
  tac: TACInstruction[];
  optimization: OptimizationDiff | null;
  assembly: AssemblyLine[];
  status: 'success' | 'failed';
  failedAtPhase?: CompilerPhase;
}

// ---- Compilation history record (Dashboard / History pages) ----
export interface CompilationRecord {
  id: string;
  projectId: string;
  timestamp: string; // ISO 8601
  status: 'success' | 'failed';
  failedAtPhase: CompilerPhase | null;
}

// ---- Grammar Library ----
export interface GrammarProduction {
  lhs: string;
  rhs: string[];
}

export interface GrammarDefinition {
  id: string;
  name: string;
  productions: GrammarProduction[];
  sampleProgram: string;
}

// ---- Project ----
export interface Project {
  id: string;
  name: string;
  createdAt: string;
  updatedAt: string;
  language: 'c-like';
}

/** Phase display metadata -- keeps labels/order consistent across the app. */
export const COMPILER_PHASES: { key: CompilerPhase; label: string }[] = [
  { key: 'lexical', label: 'Lexical Analysis' },
  { key: 'syntax', label: 'Syntax Analysis' },
  { key: 'semantic', label: 'Semantic Analysis' },
  { key: 'intermediate', label: 'Intermediate Code' },
  { key: 'optimization', label: 'Optimization' },
  { key: 'codegen', label: 'Code Generation' },
];

File: backend/app/models/compiler.py
"""
Pydantic models for the compiler pipeline.

Source of truth: SystemDesign.md Section 3 and frontend/src/types/compiler.ts.
These field names and shapes must stay in lockstep with the TypeScript
types -- any change here without a matching frontend change breaks the
contract that Architecture.md Section 8 promised (mock adapter swaps to
httpAdapter with zero component changes).
"""

from enum import Enum

from pydantic import BaseModel, Field


class TokenType(str, Enum):
    KEYWORD = "KEYWORD"
    IDENTIFIER = "IDENTIFIER"
    OPERATOR = "OPERATOR"
    LITERAL = "LITERAL"
    PUNCTUATION = "PUNCTUATION"
    COMMENT = "COMMENT"


class CompilerPhase(str, Enum):
    LEXICAL = "lexical"
    SYNTAX = "syntax"
    SEMANTIC = "semantic"
    INTERMEDIATE = "intermediate"
    OPTIMIZATION = "optimization"
    CODEGEN = "codegen"


class Token(BaseModel):
    id: str
    type: TokenType
    value: str
    line: int
    column: int


class ASTNode(BaseModel):
    id: str
    kind: str
    children: list["ASTNode"] = Field(default_factory=list)
    line: int
    metadata: dict | None = None


ASTNode.model_rebuild()


class SymbolEntry(BaseModel):
    name: str
    type: str
    scope: str
    declaredAt: int


class Severity(str, Enum):
    ERROR = "error"
    WARNING = "warning"


class SemanticDiagnostic(BaseModel):
    severity: Severity
    message: str
    line: int
    phase: CompilerPhase


class TACInstruction(BaseModel):
    id: str
    op: str
    arg1: str | None = None
    arg2: str | None = None
    result: str | None = None
    label: str | None = None


class OptimizationDiff(BaseModel):
    before: list[TACInstruction]
    after: list[TACInstruction]
    passesApplied: list[str]


class AssemblyLine(BaseModel):
    instruction: str
    operands: list[str]
    comment: str | None = None


class CompileStatus(str, Enum):
    SUCCESS = "success"
    FAILED = "failed"


class CompilationResult(BaseModel):
    tokens: list[Token]
    ast: ASTNode | None
    symbolTable: list[SymbolEntry]
    diagnostics: list[SemanticDiagnostic]
    tac: list[TACInstruction]
    optimization: OptimizationDiff | None
    assembly: list[AssemblyLine]
    status: CompileStatus
    failedAtPhase: CompilerPhase | None = None


class CompileOptions(BaseModel):
    targetOptimizations: list[str] | None = None
    stopAtPhase: str | None = None


class CompileRequest(BaseModel):
    source: str
    options: CompileOptions | None = None


class GrammarProduction(BaseModel):
    lhs: str
    rhs: list[str]


class GrammarDefinition(BaseModel):
    id: str
    name: str
    productions: list[GrammarProduction]
    sampleProgram: str


class CompilationRecord(BaseModel):
    id: str
    projectId: str
    timestamp: str
    status: CompileStatus
    failedAtPhase: CompilerPhase | None = None


class HistoryResponse(BaseModel):
    total: int
    items: list[CompilationRecord]


class ErrorDetail(BaseModel):
    code: str
    message: str
    statusCode: int


class ErrorResponse(BaseModel):
    error: ErrorDetail
